### PR TITLE
connmngr: Add DialAddr config option

### PR DESCRIPTION
**Pull request by matheusd**
This allows package users to use a dial function that receives the full
address structure instead of just the protocol family and address as
strings.

This is required for some use cases in ln. While in btcsuite this was
implemented by changing the function signature of the Dial config
option, this is done here as an extra config option to prevent having to
perform a major version bump in the connmgr module.

### CODE REVIEW

File: connmgr/connmanager.go [hunk 1]
```diff
@@ -300,7 +310,20 @@ func (cm *ConnManager) Connect(c *ConnReq) {
...
 	}
 	log.Debugf("Attempting to connect to %v", c)
-	conn, err := cm.cfg.Dial(c.Addr.Network(), c.Addr.String())
+
+	var (
+		conn net.Conn
+		err  error
+	)
+
+	if cm.cfg.Dial != nil {
```
Comment (davecgh) [inline connmgr/connmanager.go]: It already been proven in `New` that both can't be `nil`, so this is going to needless reduce the test coverage percentage with impossible to hit conditions.

I'd rather see this do:

```Go
var con net.Conn
var err error
if cm.cfg.Dial != nil {
	conn, err = cm.cfg.Dial(c.Addr.Network(), c.Addr.String())
} else {
	conn, err = cm.cfg.DialAddr(c.Addr)
}
if err != nil {
	cm.requests <- handleFailed{c, err}
} else {
	cm.requests <- handleConnected{c, conn}
}
```

File: connmgr/connmanager.go [hunk 2]
```diff
@@ -399,9 +422,12 @@ func (cm *ConnManager) Stop() {
 // New returns a new connection manager.
 // Use Start to start connecting to the network.
 func New(cfg *Config) (*ConnManager, error) {
-	if cfg.Dial == nil {
+	if (cfg.Dial == nil) && (cfg.DialAddr == nil) {
```
Comment (davecgh) [inline connmgr/connmanager.go]: No need for parenthesis.  Please remove them so it's more consistent with the rest of the code.

File: connmgr/connmanager.go [hunk 3]
```diff
@@ -399,9 +422,12 @@ func (cm *ConnManager) Stop() {
...
 		return nil, ErrDialNil
 	}
+	if (cfg.Dial != nil) && (cfg.DialAddr != nil) {
```
Comment (davecgh) [inline connmgr/connmanager.go]: Same here.

File: connmgr/connmanager_test.go [hunk 4]
```diff
@@ -68,6 +68,14 @@ func mockDialer(network, addr string) (net.Conn, error) {
 	return c, nil
 }
 
+func mockDialerAddr(addr net.Addr) (net.Conn, error) {
```
Comment (davecgh) [inline connmgr/connmanager_test.go]: Missing the comment:
```Go
// mockDialer mocks the net.Dial interface by returning a mock connection to
// the given address.
```

**Comment by matheusd** [comment]
Updated with requested review fixes.

**Pull request merged**